File: Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/elm/execution/ExpandValueSetEvaluator.java
package org.opencds.cqf.cql.engine.elm.execution;

import org.opencds.cqf.cql.engine.exception.InvalidOperatorArgument;
import org.opencds.cqf.cql.engine.execution.Context;
import org.opencds.cqf.cql.engine.runtime.ValueSet;
import org.opencds.cqf.cql.engine.terminology.TerminologyProvider;
import org.opencds.cqf.cql.engine.terminology.ValueSetInfo;

/*
expand(valueSet ValueSet) List<Code>

The ExpandValueSet function expands the given value set using the terminology provider.
*/

public class ExpandValueSetEvaluator extends org.cqframework.cql.elm.execution.ExpandValueSet {

    public static Object expand(Context context, Object valueset) {
        if (valueset == null) {
            return null;
        }

        if (valueset instanceof ValueSet) {
            TerminologyProvider tp = context.resolveTerminologyProvider();
            return tp.expand(ValueSetInfo.fromValueSet((ValueSet)valueset));
        }

        throw new InvalidOperatorArgument(
            "ExpandValueSet(ValueSet)",
            String.format("ExpandValueSet(%s)", valueset.getClass().getName())
        );
    }

    @Override
    protected Object internalEvaluate(Context context) {
        Object valueset = getOperand().evaluate(context);
        return expand(context, valueset);
    }
}

File: Src/java/engine/src/commonMain/kotlin/org/opencds/cqf/cql/engine/elm/executing/ExpandValueSetEvaluator.kt
package org.opencds.cqf.cql.engine.elm.executing

import kotlin.jvm.JvmStatic
import org.opencds.cqf.cql.engine.exception.InvalidOperatorArgument
import org.opencds.cqf.cql.engine.execution.State
import org.opencds.cqf.cql.engine.runtime.List
import org.opencds.cqf.cql.engine.runtime.Value
import org.opencds.cqf.cql.engine.runtime.ValueSet
import org.opencds.cqf.cql.engine.runtime.toCqlList
import org.opencds.cqf.cql.engine.terminology.ValueSetInfo

/*
expand(valueSet ValueSet) List<Code>

The ExpandValueSet function expands the given value set using the terminology provider.
*/
object ExpandValueSetEvaluator {
    @JvmStatic
    fun expand(valueset: Value?, state: State?): List? {
        if (valueset == null) {
            return null
        }

        if (valueset is ValueSet) {
            val tp = state!!.environment.terminologyProvider
            return tp!!.expand(ValueSetInfo.fromValueSet(valueset)).toCqlList()
        }

        throw InvalidOperatorArgument(
            "ExpandValueSet(ValueSet)",
            "ExpandValueSet(${valueset.typeAsString})",
        )
    }
}
